### PR TITLE
Implement java.lang.String toUpperCase and toLowerCase

### PR DESCRIPTION
**Pull request by ekrich**
Put together a good start to implementing `java.util.Locale`.

### CODE REVIEW

**Comment by ekrich** [comment]
This also fixes `java.nio.charset.Charset.forName` which calls `toLowerCase`.

**Comment by ekrich** [comment]
I don't think the synchronized blocks are needed for hashCode in Locale as Locale is immutable. I'm Also I am not sure about Locale.setDefault as assignment is atomic? so what would be the point?

File: javalib/src/main/scala/java/lang/String.scala [hunk 1]
```diff
@@ -633,18 +633,49 @@ final class _String()
...
-  def toUpperCase(locale: Locale): _String = ???
-
+  def toUpperCase(locale: Locale): _String = {
+    toCase(locale, Character.toUpperCase)
+  }
   def toUpperCase(): _String =
     toUpperCase(Locale.getDefault)
 
+  private[this] def toCase(locale: Locale, convert: Int => Int): _String = {
+    if (count == 0) value
```
Comment (MasseGuillaume) [inline javalib/src/main/scala/java/lang/String.scala]: it ignores the locale for now?

Comment (ekrich) [inline javalib/src/main/scala/java/lang/String.scala]: In the Harmony 6.0 there was a special case for Turkish. The newer Harmony https://github.com/apache/harmony/blob/java6/classlib/modules/luni/src/main/java/java/lang/String.java defers to icu4j `UCharacter` which defers to UCaseProps which has a couple of special cases for Turkish and Lithuanian but I think this should gets us pretty close. Overall, this stuff is really complex especially icu4j which I guess is a complete Unicode implementation.

File: javalib/src/main/scala/java/util/Locale.scala [hunk 2]
```diff
@@ -1,7 +1,165 @@
...
+
+  //def getUnicodeLocaleAttributes(): ju.Set[String] = ???
+
+  //def getUnicodeLocaleKeys(): ju.Set[String] = ???
+
+  //def getUnicodeLocaleType(key: String) = ???
+
+  //def getVariant(): String = variant
+
+  // synchronized in Java
```
Comment (densh) [inline javalib/src/main/scala/java/util/Locale.scala]: This comment is not very information, lets keep it out.

File: javalib/src/main/scala/java/util/Locale.scala [hunk 3]
```diff
@@ -1,7 +1,165 @@
...
+  @inline override def hashCode(): Int = this.synchronized {
+    country.hashCode() + language.hashCode() + variant.hashCode()
+  }
+
+  //def toLanguageTag(): String = ???
+
+  // Examples: "en", "en_US", "_US", "en__POSIX", "en_US_POSIX"
+  @inline override def toString(): String = {
+    // straight port
```
Comment (densh) [inline javalib/src/main/scala/java/util/Locale.scala]: This one too should probably not be there.

File: javalib/src/main/scala/java/lang/String.scala (hunk 1 above)
Comment (densh) [inline javalib/src/main/scala/java/lang/String.scala]: I think it's acceptable to ignore it for now as long as there is an issue open, we'll improve it later on.

File: javalib/src/main/scala/java/lang/String.scala [hunk 4]
```diff
@@ -633,18 +633,52 @@ final class _String()
...
   def toLowerCase(): _String =
     toLowerCase(Locale.getDefault)
 
   override def toString(): String = this
 
-  def toUpperCase(locale: Locale): _String = ???
-
+  def toUpperCase(locale: Locale): _String = {
+    toCase(locale, Character.toUpperCase)
+  }
```
Comment (densh) [inline javalib/src/main/scala/java/lang/String.scala]: Redundant curly braces around the method body.

File: javalib/src/main/scala/java/lang/String.scala [hunk 5]
```diff
@@ -633,18 +633,52 @@ final class _String()
     buffer
   }
 
-  def toLowerCase(locale: Locale): _String = ???
+  def toLowerCase(locale: Locale): _String = {
+    toCase(locale, Character.toLowerCase)
+  }
```
Comment (densh) [inline javalib/src/main/scala/java/lang/String.scala]: Redundant curly braces around the method body.

File: javalib/src/main/scala/java/lang/String.scala [hunk 6]
```diff
@@ -633,18 +633,52 @@ final class _String()
...
-  def toUpperCase(locale: Locale): _String = ???
-
+  def toUpperCase(locale: Locale): _String = {
+    toCase(locale, Character.toUpperCase)
+  }
   def toUpperCase(): _String =
     toUpperCase(Locale.getDefault)
 
+  private[this] def toCase(locale: Locale, convert: Int => Int): _String = {
+    if (count == 0) value
```
Comment (densh) [inline javalib/src/main/scala/java/lang/String.scala]: Did you mean `if (count == 0) return value` ? Currently this check doesn't do anything.

File: javalib/src/main/scala/java/util/Locale.scala [hunk 7]
```diff
@@ -1,7 +1,165 @@
 package java.util
 
-final class Locale
+//import java.{util => ju}
```
Comment (densh) [inline javalib/src/main/scala/java/util/Locale.scala]: This line is not necessary.

File: javalib/src/main/scala/java/lang/String.scala (hunk 6 above)
Comment (ekrich) [inline javalib/src/main/scala/java/lang/String.scala]: My intent here was to handle the empty String case - yes that is what I meant.

File: javalib/src/main/scala/java/util/Locale.scala (hunk 7 above)
Comment (ekrich) [inline javalib/src/main/scala/java/util/Locale.scala]: I left that line commented out as other commented methods return `ju.Set`. Other places in the code used that alias so I thought it was helpful.

Comment (densh) [inline javalib/src/main/scala/java/util/Locale.scala]: I see, lets keep it for now then.

**Comment by densh** [comment]
LGTM, thanks @ekrich ! ✨🍰 ✨

**Pull request merged**